### DISCUSSION TITLE
Allow to implement your own authentication method

### DISCUSSION THREAD

**Discussion by hardening** [curl/curl#11399] [Ideas]
Hi, 
I'm working on the FreeRDP project, and we would be interested to use `curl` in the RDP scenarii that involve HTTP (replace our custom HTTP and websocket methods by curl ones).
Anyway for that, we'd like to be able to wire our authentication routines, as we want to control things like `NTLM`, `Negotiate` ourselves as most of the time, system's libraries don't provide what we want in terms of features or configuration.

Would it be possible to have an API to wire our own routines and functions that could control the behavior of the authentication workflow (existing code for auth methods could be just special cases of this general API) ?

**Comment by bagder** [comment]
> Hi, I'm working on the FreeRDP project, and we would be interested to use `curl` in the RDP scenarii that involve HTTP (replace our custom HTTP and websocket methods by curl ones). Anyway for that, we'd like to be able to wire our authentication routines, as we want to control things like `NTLM`, `Negotiate` ourselves as most of the time, system's libraries don't provide what we want in terms of features or configuration.
> 
> Would it be possible to have an API to wire our own routines and functions that could control the behavior of the authentication workflow (existing code for auth methods could be just special cases of this general API) ?

Sure it would. I think the hardest part of that job is to design such an API that you can understand how to use it and yet features the sufficient amount of flexibility and power.

I would imagine that the best way to design the API is to also create an example application that uses it in a manner similar to the use case you have in mind.

**Reply by hardening** [reply]
Another possibility would be to design an API that can hold existing authentication methods in curl, like for instance when you set basic auth that would install the basic Auth callbacks. And I suppose that when you've ported all the existing auth methods that are already there, then the API is able to treat all the possible cases (because you already covers a quite large range with the existing auth methods).

**Comment by dfandrich** [comment]
Would you actually need any special support for this? Could not not simply not
turn on any libcurl authentication methods, perform a request, then parse the
resulting headers yourself? I can't imagine an authentication API that would be
sufficiently flexible to handle any authentication case that could be dreamed
up that would provide much more than that, anyway.

**Reply by hardening** [reply]
Hum yeah perhaps that would be the easiest for us, can libcurl already do that ? I feel like it looks at the answers and do special treatments when the response is 401...

**Comment by dfandrich** [comment]
Yes, it can do that already, with your choice of ways to get and set headers.
